### DISCUSSION TITLE
Is there a way to make deep copies?

### DISCUSSION THREAD

**Discussion by ozaner** [k-paxian/dart-json-mapper#208] [Q&A] [answered]
The [objects cloning section](https://github.com/k-paxian/dart-json-mapper#objects-cloning) in the readme show how to make a copy of a mappable object with one of its properties changed. Is it possible to do that with nested properties? freezed, for example, allows cloning with changes in a lower level property:
```dart
final cloneCar = car.copyWith.leftWheel.tire(thickness: 32);
```

But dart_json_mappable's copyWith method only allows for one deep changes:
```dart
final cloneCar = JsonMapper.copyWith(car, {'color': 'blue'}); // same object but blue
```

If we were to change a deeper property like so:
```dart
final cloneCar = JsonMapper.copyWith(car, {'leftWheel': {'tire': {'thickness': 32}}}); // replaces entire leftWheel
```

It would replace the entirety of the top level property. If I was able to deep clone objects like above, then I could use mappable classes as immutable data classes. Is there any way to achieve this?

**Comment by k-paxian** [comment, marked as answer]
Sure, it's possible. Please refer [this ](https://github.com/k-paxian/dart-json-mapper/commit/f045498dc38f7a69827aa1987d5992f446181880) commit for a unit test demonstrating exactly that.

**Reply by ozaner** [reply]
Ah, I think the reason it isn't working for me is that I'm using nested keys:

```dart
@jsonSerializable
class ComposableCar {
  @JsonProperty(name: 'leftWheel/tire/thickness')
  final int thicknessLeft;
  @JsonProperty(name: 'rightWheel/tire/thickness')
  final int thicknessRight;
  @JsonProperty(name: 'rightWheel/tire/color')
  final String? tireRightColor;
  const ComposableCar(this.thicknessLeft, this.thicknessRight, this.tireRightColor);
}
```
If we run the copyWith method on this:

```dart
var newCar = JsonMapper.copyWith(car, {'rightWheel': {'tire': {'thickness': 32}}})
```

it doesn't copy tireRightColor when creating the new car. I guess the real question is, is there a way to change a property with _nested keys_ using the copyWith method?

**Reply by k-paxian** [reply]
Seems like a bug indeed 😃
Needs proper JSON objects merging to achieve that.

**Reply by k-paxian** [reply]
Fixed  https://github.com/k-paxian/dart-json-mapper/issues/209

**Reply by ozaner** [reply]
This probably works but I can't really test it since dart_json_mapper ^2.2.7+3 is not compatible with flutter due to versioning conflicts with some packages (e.g. flutter uses collection 1.16.0 but dart_json_mapper uses collection ^1.17.0).

**Reply by k-paxian** [reply]
I hear you @ozaner  [2.2.7+4](https://github.com/k-paxian/dart-json-mapper/commit/e6924c9a78e2a1c0d67f092d42ed767c8bdcecd1)

**Reply by ozaner** [reply]
Thanks, but I'm not sure this works. Just to test, I created a fresh flutter project and added ran `flutter pub add dart_json_mapper ` and the latest version that could be resolved was `^2.2.7`.  If I were to manually change that to `^2.2.7+4` in the pubspec.yaml, I get the following error message:

```
[flutter_application_1] flutter pub get
Running "flutter pub get" in flutter_application_1...           
Because every version of flutter_test from sdk depends on path 1.8.2 and dart_json_mapper >=2.2.7+2 depends on path ^1.8.3, flutter_test from sdk is incompatible with dart_json_mapper >=2.2.7+2.
So, because flutter_application_1 depends on both dart_json_mapper ^2.2.7+4 and flutter_test from sdk, version solving failed.
pub get failed (1; So, because flutter_application_1 depends on both dart_json_mapper ^2.2.7+4 and flutter_test from sdk, version solving failed.)
exit code 1
```

And I'm not sure path being incompatible with flutter_test is the only problem, because when I comment out flutter_test from the dev_dependencies I get another error:
```
[flutter_application_1] flutter pub get
Running "flutter pub get" in flutter_application_1...           
Because dart_json_mapper >=2.2.7+2 depends on analyzer ^5.3.1 which depends on collection ^1.17.0, dart_json_mapper >=2.2.7+2 requires collection ^1.17.0.
And because every version of flutter from sdk depends on collection 1.16.0, dart_json_mapper >=2.2.7+2 is incompatible with flutter from sdk.
So, because flutter_application_1 depends on both flutter from sdk and dart_json_mapper ^2.2.7+4, version solving failed.
pub get failed (1; So, because flutter_application_1 depends on both flutter from sdk and dart_json_mapper ^2.2.7+4, version solving failed.)
exit code 1
```

Which is strange because I see that you downgraded the collection dependency to 1.16.0 in 2.2.7+4.

**Reply by k-paxian** [reply]
ok, `^2.2.7+5` downgraded `path` as well @ozaner hope this helps 🤞
